// @storybook/addon-storysource/dist/preview.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withStorySource = withStorySource;

var _addons = require('@storybook/addons');

var _addons2 = _interopRequireDefault(_addons);

var _ = require('./');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getLocation(context, locationsMap) {
  return locationsMap[context.kind + '@' + context.story] || locationsMap['@' + context.story];
}

function setStorySource(context, source, locationsMap) {
  var channel = _addons2.default.getChannel();
  var currentLocation = getLocation(context, locationsMap);

  channel.emit(_.EVENT_ID, {
    source: source,
    currentLocation: currentLocation,
    locationsMap: locationsMap
  });
}

function withStorySource(source) {
  var locationsMap = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  return function (story, context) {
    setStorySource(context, source, locationsMap);
    return story();
  };
}
