// @storybook/addon-storysource/dist/loader/inject-decorator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defaultOptions = require('./default-options');

var _defaultOptions2 = _interopRequireDefault(_defaultOptions);

var _generateHelpers = require('./generate-helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function extendOptions(source, comments, options) {
  return (0, _extends3.default)({}, _defaultOptions2.default, options, {
    source: source,
    comments: comments
  });
}

function inject(source, decorator) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  var _generateSourceWithDe = (0, _generateHelpers.generateSourceWithDecorators)(source, decorator),
      changed = _generateSourceWithDe.changed,
      newSource = _generateSourceWithDe.source,
      comments = _generateSourceWithDe.comments;

  if (!changed) {
    return {
      source: newSource,
      addsMap: {},
      changed: changed
    };
  }

  var storySource = (0, _generateHelpers.generateStorySource)(extendOptions(source, comments, options));
  var addsMap = (0, _generateHelpers.generateAddsMap)(storySource);

  return {
    source: newSource,
    storySource: storySource,
    addsMap: addsMap,
    changed: changed
  };
}

exports.default = inject;
